packages/crypto/src/Bip32/Bip32KeyDerivation.ts

```diff
@@ -0,0 +1,175 @@
+import { BN } from 'bn.js';
+import { InvalidArgumentError } from '../errors';
+import { crypto_auth_hmacsha512, crypto_scalarmult_ed25519_base_noclamp } from 'libsodium-wrappers-sumo';
+import { eddsa } from 'elliptic';
+
+const ed25519 = new eddsa('ed25519');
+
+/**
+ * Check if the index is hardened.
+ *
+ * @param index The index to verify.
+ * @returns true if hardened; otherwise; false.
+ */
+const isHardenedDerivation = (index: number) => index >= 0x80_00_00_00;
+
+/**
+ * Derives the private key with a hardened index.
+ *
+ * @param index The derivation index.
+ * @param scalar Ed25519 curve scalar.
+ * @param iv Ed25519 binary blob used as IV for signing.
+ * @param chainCode The chain code.
+ */
+const deriveHardened = (
+  index: number,
+  scalar: Buffer,
+  iv: Buffer,
+  chainCode: Buffer
+): { zMac: Uint8Array; ccMac: Uint8Array } => {
+  const data = Buffer.allocUnsafe(1 + 64 + 4);
+  data.writeUInt32LE(index, 1 + 64);
+  scalar.copy(data, 1);
+  iv.copy(data, 1 + 32);
+
+  data[0] = 0x00;
+  const zMac = crypto_auth_hmacsha512(data, chainCode);
+  data[0] = 0x01;
+  const ccMac = crypto_auth_hmacsha512(data, chainCode);
+
+  return { ccMac, zMac };
+};
+
+/**
+ * Derives the private key with a 'soft' index.
+ *
+ * @param index The derivation index.
+ * @param scalar Ed25519 curve scalar.
+ * @param chainCode The chain code.
+ */
+const deriveSoft = (index: number, scalar: Buffer, chainCode: Buffer): { zMac: Uint8Array; ccMac: Uint8Array } => {
+  const data = Buffer.allocUnsafe(1 + 32 + 4);
+  data.writeUInt32LE(index, 1 + 32);
+
+  const vk = Buffer.from(crypto_scalarmult_ed25519_base_noclamp(scalar));
```

Comment:
Just a thought here, but it might be more efficient to store the VK together with the secret key, rather than compute it each time that is needed.

Reply:
The savings are not massive (though it is one of the most expensive cryptographic operations). The reason why I said this was because I saw the use of `crypto_scalarmult_ed25519_base_noclamp` in the signature function as well, and thought that it could save some computation to simply store it as part of the extended private key for instance. However, performance does not seem to be paramount to the point of wanting to reduce 50us, so it should be ok.